scheduler/core/components/optimizer/greedymax.py

```diff
@@ -15,18 +15,45 @@
 from .base import BaseOptimizer
 
 import numpy as np
-import astropy.units as u
+# import astropy.units as u
 
 
 class GreedyMaxOptimizer(BaseOptimizer):
     """
     GreedyMax is an optimizer that schedules the visits for the rest of the night in a greedy fashion.
     """
 
-    def __init__(self, seed=42):
+    def __init__(self):
         self.groups = []
-        self.groupids = []
+        self.group_ids = []
+        # self.obs_groups = []     # remove if not used
+        self.obs_group_ids = []
         self.timelines = []
+        self.sites = []
+        self.min_visit_len = timedelta(minutes=30)
+
+    def setup(self, selection: Selection) -> GreedyMaxOptimizer:
+        """
+        Preparation for the optimizer e.g. create chromosomes, etc.
+        """
+        self.groups = list(selection.schedulable_groups.values())
+        self.group_ids = list(selection.schedulable_groups)
+        for gid, group_data in selection.schedulable_groups.items():
+            if group_data.group.is_observation_group():
+                self.obs_group_ids.append(group_data.group.unique_id())
+                # self.obs_groups.append(group_data.group)
+            elif group_data.group.is_scheduling_group():
+                for subgroup in group_data.group.children:
+                    if subgroup.is_observation_group():
+                        self.obs_group_ids.append(subgroup.unique_id())
+                        # self.obs_groups.append(subgroup)
+
+        period = len(list(selection.night_events.values())[0].time_grid)
+        print('Period: ', period)
+        self.timelines = [Timelines(selection.night_events, night) for night in range(period)]
+        self.sites = list(selection.night_events.keys())
```

Comment:
That would be good, though we should rename 'period' to something descriptive like 'number_nights_plan'

Reply:
I'll make a Jira task to do this and we can include that right away since it's just a few minutes of work.

Let me know if there's some information that you have to get at in a difficult way like this: it's probably indicative that we should add it to the model data being passed to the code using it.

Reply:
Okay, the `Selection` will have the following data:
* `num_nights_plan`, which returns an `int`
* `sites`, which returns a `FrozenSet[Site]`.